dwh/indicator_lib.py

```diff
@@ -74,65 +71,22 @@ def calc_TX_PVLS(patient_agg_obs: pandas.DataFrame, VL_code: str,
 
   Args:
-    patient_agg_obs: A DataFrame generated by `join_patients_agg_obs()`.
+    patient_agg_obs: An output from `patient_query.find_patient_aggregates()`.
     VL_code: The code for viral load values.
     failure_threshold: VL count threshold of failure.
     end_date: The string representation of the last date as 'YYYY-MM-DD'.
   Returns:
+    The aggregated DataFrame with age/gender buckets.
   """
   end_date = datetime.today()
   if end_date_str:
     end_date = datetime.strptime(end_date_str, '%Y-%m-%d')
-  agg_buckets_udf = F.UserDefinedFunction(
-      lambda a, g: _agg_buckets(a, g, end_date),
-      T.ArrayType(T.StringType()))
-  VL_df = patient_agg_obs.withColumn(
-      'sup_VL', patient_agg_obs[VL_code + '_max_value'] < failure_threshold).withColumn(
-      'agg_buckets', agg_buckets_udf(
-          patient_agg_obs['birthDate'], patient_agg_obs['gender'])
-  )
-  num_patients = VL_df.count()
-  VL_agg_P = VL_df.select(
-      VL_df.sup_VL,
-      F.explode(VL_df._agg_buckets).alias('agg_bucket')).groupBy(
-      'sup_VL', 'agg_bucket').agg(
-      F.count('*').alias('count')).toPandas().sort_values(
-      ['agg_bucket', 'sup_VL'])
-  VL_agg_P['ratio'] = VL_agg_P['count']/num_patients
-  return VL_agg_P
-
-
-# TODO remove this function once we verify that that pandas.DataFrame based
-#   approach scales.
-#def calc_TX_PVLS_spark(patient_agg_obs: DataFrame, VL_code: str, failure_threshold: int,
-#    end_date_str: str = None) -> pandas.DataFrame:
-#  """Calculates TX_PVLS indicator with its corresponding disaggregations.
-#
-#  Args:
-#    patient_agg_obs: A DataFrame generated by `join_patients_agg_obs()`.
-#    VL_code: The code for viral load values.
-#    failure_threshold: VL count threshold of failure.
-#    end_date: The string representation of the last date as 'YYYY-MM-DD'.
-#  Returns:
-#  """
-#  end_date = datetime.today()
-#  if end_date_str:
-#    end_date = datetime.strptime(end_date_str, '%Y-%m-%d')
-#  agg_buckets_udf = F.UserDefinedFunction(
-#      lambda a, g: _agg_buckets(a, g, end_date),
-#      T.ArrayType(T.StringType()))
-#  VL_df = patient_agg_obs.withColumn(
-#      'sup_VL', patient_agg_obs[VL_code + '_max_value'] < failure_threshold).withColumn(
-#      'agg_buckets', agg_buckets_udf(
-#          patient_agg_obs['birthDate'], patient_agg_obs['gender'])
-#  )
-#  num_patients = VL_df.count()
-#  VL_agg_P = VL_df.select(
-#      VL_df.sup_VL,
-#      F.explode(VL_df._agg_buckets).alias('agg_bucket')).groupBy(
-#      'sup_VL', 'agg_bucket').agg(
-#      F.count('*').alias('count')).toPandas().sort_values(
-#      ['agg_bucket', 'sup_VL'])
-#  VL_agg_P['ratio'] = VL_agg_P['count']/num_patients
-#  return VL_agg_P
-
+  temp_df = patient_agg_obs[(patient_agg_obs['code'] == VL_code)].copy()
+  # Note the above copy is used to avoid setting a new column on a slice next:
+  temp_df['sup_VL'] = (temp_df['max_value'] < failure_threshold)
```

Comment:
I am still worried about min_value and max_value.  Suppose a patient has 2 VLs (100,003 and 54) within the reporting period.

Using the max_value to calculate  the TX_PVLS would make this patient un-suppressed since the latest VL is 54.

Instead, it would be better to have `current_value` and ` previous_value`, that way we can be able to use relevant values within a given reporting period.

And please to correct me if this is already, or can be handled?

![Screenshot 2021-06-24 at 10 21 42 AM](https://user-images.githubusercontent.com/10988005/123219807-fba5c600-d4d5-11eb-90ff-f4115136ee9c.png)

Reply:
Okay per your comments here and also on chat, I have added a few more columns in the returned DataFrame. In particular please take a look at the new `{min,max}_date_valye_*` columns. I have a feeling that this is not the right way of doing this and Spark should have some built-in constructs to do this but I have not found one. My approach seems to have some non-trivial performance impact too but I have not focused on performance for now, anyways.

Please take a look at the new commit and let me know what you think.

Reply:
Great! This looks fine, let's worry about the performance later.  Thanks @bashir2 for working on